### PR TITLE
Fix flaky stateless test with azure

### PR DESCRIPTION
**Pull request by SmitaRKulkarni**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Reduce the memory usage when using Azure object storage by using fixed memory allocation, avoiding the allocation of an extra buffer.

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [x]  Fast test
- [x]  Integration Tests
- [x]  Stateless tests
- [x]  Stateful tests
- [x]  Unit tests
- [x]  Performance tests
- [x]  All with ASAN
- [x]  All with TSAN
- [x]  All with Analyzer
- [x]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 7aaf7e0e03d1d1e192a68277c4f6836ce92587e4 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63160/7aaf7e0e03d1d1e192a68277c4f6836ce92587e4/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by SmitaRKulkarni** [comment]
Updated WriteBufferFromAzureBlobStorage to use required memory, instead of allocating extra buffer (just like we do for s3 : https://github.com/ClickHouse/ClickHouse/pull/44869/files)

File: src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp [hunk 1]
```diff
@@ -119,22 +123,32 @@ void WriteBufferFromAzureBlobStorage::preFinalize()
...
+    setFakeBufferWhenPreFinalized();
+
     /// If there is only one block and size is less than or equal to max_single_part_upload_size
     /// then we use single part upload instead of multi part upload
-    if (buffer_allocation_policy->getBufferNumber() == 1)
+    if (detached_part_data.size() == 1 && block_ids.empty())
     {
-        size_t data_size = size_t(position() - memory.data());
-        if (data_size <= max_single_part_upload_size)
+        if (detached_part_data.front().data_size <= max_single_part_upload_size)
```
Comment (jkartseva) [inline src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp:137]: Can be combined for readability:
``` cpp
if (block_ids.empty() && detached_part_data.size() == 1 && detached_part_data.front().data_size <= max_single_part_upload_size)
{
    // a single-part upload
}
else
{
    writeMultipartUpload();
}
```

**Review by jkartseva — APPROVED** [review]
Please take a look at the comments.

File: src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp [hunk 2]
```diff
@@ -119,22 +123,32 @@ void WriteBufferFromAzureBlobStorage::preFinalize()
...
+            auto part_data = std::move(detached_part_data.front());
             auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
-            Azure::Core::IO::MemoryBodyStream memory_stream(reinterpret_cast<const uint8_t *>(memory.data()), data_size);
-            execWithRetry([&](){ block_blob_client.Upload(memory_stream); }, max_unexpected_write_error_retries, data_size);
+            Azure::Core::IO::MemoryBodyStream memory_stream(reinterpret_cast<const uint8_t *>(part_data.memory.data()), part_data.data_size);
+            execWithRetry([&](){ block_blob_client.Upload(memory_stream); }, max_unexpected_write_error_retries, part_data.data_size);
             LOG_TRACE(log, "Committed single block for blob `{}`", blob_path);
+
+            detached_part_data.pop_front();
+            block_ids.clear();
```
[the remark this answers is not in the document]
Comment (SmitaRKulkarni) [reply src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp:146]: removed block_ids.clear();

File: src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp (hunk 1 above)
Comment (SmitaRKulkarni) [reply src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp:137]: addressed it

File: src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp [hunk 3]
```diff
@@ -172,30 +242,57 @@ void WriteBufferFromAzureBlobStorage::allocateBuffer()
...
+    {
+        writePart(std::move(detached_part_data.front()));
+        detached_part_data.pop_front();
+    }
+}
+
 }
 
 #endif
+
```
[the remark this answers is not in the document]
Comment (SmitaRKulkarni) [reply src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp:298]: addressed it

File: src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp [hunk 4]
```diff
@@ -156,13 +169,60 @@ void WriteBufferFromAzureBlobStorage::finalizeImpl()
...
+    if (memory.size() == max_first_buffer)
+        return;
+
+    size_t size = std::min(memory.size() * 2, max_first_buffer);
+    memory.resize(size);
+
+    WriteBuffer::set(memory.data() + hidden_size, memory.size() - hidden_size);
+    chassert(offset() == 0);
+}
+
```
[the remark this answers is not in the document]
Comment (SmitaRKulkarni) [reply src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp:228]: agreed, would be good to combine some functions and have a class such as WriteBufferFromObjectStorageUtils, will do a follow-up PR

**Review by jkartseva — APPROVED** [review]
Looks good. Test failure seems irrelevant https://s3.amazonaws.com/clickhouse-test-reports/63160/9ce824d41c2323f6556dd6f4015d9ec99272083c/stateless_tests_flaky_check__asan_.html

**Comment by SmitaRKulkarni** [comment]
This failure is also not related 
Integration Tests (asan, Old Analyzer) [2/6] - test_mask_sensitive_info/test.py::test_create_table - will be fixed by #64901

**Pull request merged**